src/v/wasm/wasmtime.cc

```diff
@@ -540,20 +555,158 @@ class wasmtime_engine : public engine {
+        }
+        co_return std::move(fut).get();
+    }
+
+private:
+    ss::future<> initialize_wasi() {
+        auto* ctx = wasmtime_store_context(_store.get());
+        wasmtime_extern_t start;
+        bool ok = wasmtime_instance_export_get(
+          ctx,
+          &_instance,
+          wasi::preview_1_start_function_name.data(),
+          wasi::preview_1_start_function_name.size(),
+          &start);
+        if (!ok || start.kind != WASMTIME_EXTERN_FUNC) {
+            throw wasm_exception(
+              "Missing wasi _start function", errc::user_code_failure);
+        }
+        vlog(wasm_log.info, "starting wasm vm {}", _meta.name());
+        std::exception_ptr ex;
+        try {
+            co_await call_host_func(ctx, &start.of.func, {}, {});
+        } catch (...) {
+            ex = std::current_exception();
+            vlog(wasm_log.warn, "wasm vm failed: {}", ex);
+        }
+        vlog(wasm_log.info, "wasm vm {} finished", _meta.name());
+        if (!ex) {
+            ex = std::make_exception_ptr(
+              wasm_exception("vm has exited", errc::engine_not_running));
+        }
+        _transform_module_v2.stop(ex);
+    }
+
+    ss::future<model::record_batch>
+    invoke_transform(model::record_batch batch, transform_probe* p) {
+        std::unique_ptr<transform_probe::hist_t::measurement> m;
+        model::record_batch_header header = batch.header().copy();
+        auto transformed = co_await _transform_module_v2.for_each_record_async(
+          std::move(batch),
+          [this, &m, p, ctx = wasmtime_store_context(_store.get())]() {
+              reset_fuel(ctx);
+              m = p->latency_measurement();
+          });
+        m = nullptr;
+
+        model::record_batch::compressed_records serialized_records;
+        int32_t i = 0;
+        for (model::transformed_data& r : transformed) {
+            serialized_records.append_fragments(
+              std::move(r).to_serialized_record(
+                model::record_attributes(),
+                /*timestamp_delta=*/0,
+                /*offset_delta=*/i++));
+        }
+
+        header.record_count = i;
+        header.size_bytes = int32_t(
+          model::packed_record_batch_header_size
+          + serialized_records.size_bytes());
+
+        // mark the batch as created with broker time
+        header.attrs.set_timestamp_type(model::timestamp_type::append_time);
+        header.first_timestamp = model::timestamp::now();
+        header.max_timestamp = header.first_timestamp;
+
+        batch = model::record_batch(
+          header,
+          std::move(serialized_records),
+          model::record_batch::tag_ctor_ng{});
+
+        // Recompute the crc
+        batch.header().crc = model::crc_record_batch(batch);
+        batch.header().header_crc = model::internal_header_only_crc(
+          batch.header());
+
+        co_return batch;
+    }
+
+    ss::future<> _main_task = ss::now();
```

Comment:
looks sus

Reply:
Can you elaborate? Which part?

Reply:
nevermind, i thought we were in function scope here on this line and you were accidentally shadowing a class variable named `_main_task`.

Reply:
😮‍💨